### ISSUE TITLE
Rows cannot implement Drop

### ISSUE THREAD

**Issue by gwenn** [jgallagher/rusqlite#150]
```rust
let row = s.query(&[]).unwrap().next().unwrap().unwrap();
// Rows has already been dropped, statement has been reset, no data can be retrieved...
row.get(0)
```

**Comment by gwenn** [comment]
See https://github.com/gwenn/rusqlite/commit/d8d220e2dd97ea10ce0c46c675e4991f58c34be4

**Comment by jgallagher** [comment]
Ick. Great catch. I would like for that to not compile, but AFAICT an `Iterator` impl can't return references to itself from `next`, which is what I believe we'd need.

Options:

* Leave it as is and document it. This is related to the already-documented issue that if you call `next` twice, you'll get errors trying to access the first row which is now stale.
* Don't reset when `Rows` is dropped, and go back to `reset_if_needed`. We can keep the `reset` on error / `SQLITE_DONE` to still reset ASAP when possible.

Thoughts? Other options?

**Comment by jgallagher** [comment]
Another option is to deprecate/remove `query`. We already recommend people use `query_map` and/or `query_and_then` instead, neither of which expose this problem.

**Comment by gwenn** [comment]
Could you please confirm that the problem is related to [streaming iterators](https://www.reddit.com/r/rust/comments/467j1p/iterator_with_item_living_for_one_iteration) ?

**Comment by jgallagher** [comment]
Yes, that's the exact problem.

**Comment by jgallagher** [comment]
I think #153 is a decent compromise. Keep `Rows` but change `next()` to return a `Row<'a, 'stmt>` where `'a` is tied to the vending `Rows` handle. This means `Rows` no longer implements `Iterator`, but it's at least still usable, and this change doesn't affect anyone using `query_map`/`query_and_then`.

**Comment by gwenn** [comment]
This change breaks the current API (when users iterate on Rows), no ?

**Comment by jgallagher** [comment]
It does, yeah.

We're still in `0.x` versioning, so breaking changes are allowed between minor versions. I was okay with runtime checks to let `Rows` implement `Iterator` when the only thing that could go wrong was if you used a row after calling `next()` again, but adding more runtime checks (or disabling resetting the statement when a `Rows` is dropped) seems bad.

**Comment by gwenn** [comment]
Ok.

**Comment by jgallagher** [comment]
Closed with #153.
